### DISCUSSION TITLE
Seagate ST11 V2.0 BIOS + Floppy drive ROMs

### DISCUSSION THREAD

**Discussion by 640-KB** [640-KB/GLaBIOS#37] [Tech Support] [answered]

**Comment by xinteksik** [comment]
I was trying to use "floppy bios", but no success.

https://github.com/640-KB/GLaBIOS/wiki/High-Density-Floppy

I use 64k EEPROM (last 8k is GlaBIOS start at E000 and at C000 is floppy bios). PC starts fine, GlaBIOS show info and than strart Floppy Bios. But still is showing "...pres F2 to setup..." again and again (in loop). Reconfiguring Floppy bios to actual HW setup (one floppy 1,2M) nothing change.

Is there any guide how to get it works toghether?

XT clone MAXAR 88L (FE2010A + 8088 + 8087).

**Comment by TzOk83** [comment]
Try to use some HD FDD controller BIOS from the era... I found two HD BIOSes, DTK and JOINCOM ELECTRONIC CORP. working with GLaBIOS.

**Comment by xinteksik** [comment]
> Try to use some HD FDD controller BIOS from the era... I found two HD BIOSes, DTK and JOINCOM ELECTRONIC CORP. working with GLaBIOS.

I was trying step by step all floppy controller bioses from (https://www.minuszerodegrees.net/rom/rom.htm and https://lo-tech.co.uk/wiki/Lo-tech_ISA_ROM_Board) and no succes.
Floppy controller on my board is WD37C65B (is nearest to connector). Some bioses was loaded, some not.

![tempImagehSi9xi](https://github.com/640-KB/GLaBIOS/assets/12784037/c1026439-e0f4-4657-a5c4-013992ded3a0)

**Comment by 640-KB** [comment]
Based on the looping and the fact that the DTK is loaded twice it almost seems like the floppy BIOS is getting loaded twice - like it's duplicated in the ROM at two different offsets?

In the latest (not fully released) version of GLaBIOS there's a feature to show the offset of a ROM as it's being loaded which would tell us if that theory is right or not. I can make you a build of that a little bit later today.

Also, with the Multi Floppy BIOS the F2 configuration only works directly if you have a card with a AT28C64B EEPROM (Sergey Kiselev's hardware commonly). Otherwise it will tell you which bytes to edit in the rom image for the drive config you have.

**Comment by 640-KB** [comment]
@xinteksik Give this version of GLaBIOS a try - it should help us troubleshoot this a little bit and figure out what's going on!

[GLABIOS_0.2.5_8F_072723_1.zip](https://github.com/640-KB/GLaBIOS/files/12186448/GLABIOS_0.2.5_8F_072723_1.zip)

**Comment by xinteksik** [comment]
Thanks a lot

> Based on the looping and the fact that the DTK is loaded twice it almost seems like the floppy BIOS is getting loaded twice - like it's duplicated in the ROM at two different offsets?

I upload ROM image (64k). At E000 is glabios, and at C000 is fdd. Nothing else.  [bios.zip](https://github.com/640-KB/GLaBIOS/files/12187557/bios.zip)

> In the latest (not fully released) version of GLaBIOS there's a feature to show the offset of a ROM as it's being loaded which would tell us if that theory is right or not. I can make you a build of that a little bit later today.

I upload photo, but nothing is show.
![IMG_1987](https://github.com/640-KB/GLaBIOS/assets/12784037/7890df54-acc1-4b71-8135-f2df9cc90bb6)


> Also, with the Multi Floppy BIOS the F2 configuration only works directly if you have a card with a AT28C64B EEPROM (Sergey Kiselev's hardware commonly). Otherwise it will tell you which bytes to edit in the rom image for the drive config you have.

Yes, I edit rom image and write again.

![image](https://github.com/640-KB/GLaBIOS/assets/12784037/39c31037-c41e-4845-8e12-0adb6b29dc4b)

Partialy working was glabios (from previous post) + FDC-BIOS.ROM [FDC-BIOS.zip](https://github.com/640-KB/GLaBIOS/files/12187601/FDC-BIOS.zip) Starting MS-DOS was from 1,2MB floppy (show as 1,44), but hangs.
![IMG_1980](https://github.com/640-KB/GLaBIOS/assets/12784037/832b0c55-edc1-421d-a6d8-b0ff91ce6762)

**Comment by Maxx1234** [comment]
Some similar experience I had with XT-IDE BIOS and Sergey's Floppy BIOS.

1.Sergey's Floppy BIOS versions >2.4 do not work correctly on my ST-12 and DTK clone systems.  This may be due to modification of the IPL code.
2.XT-IDE ROM starting address must be higher than Sergey's Floppy BIOS ROM starting address. If not, boot sequence is handled by the last ROM loaded (Floppy) and the system cannot boot from the hard disk.
3.In order to configure the Floppy BIOS on a live system (using an AT28C64, pressing F2 at boot), XT-IDE BIOS should not be present.  After saving the configuration, XT-IDE BIOS can be enabled and the boot sequence works fine.
4.XT-IDE BIOS allows manually booting from a FDD (using drive letter hotkey), simplifying the boot process. Can't tell how the ST11 BIOS handles the boot sequence, if loaded last (higher starting address than Floppy BIOS).

Good luck!

**Comment by 640-KB** [comment]
@xinteksik Oops, looks like the build I sent didn't have the option ROM display enabled after all.  This one is the right one. :)
[GLABIOS_0.2.5_8F_072723_2.zip](https://github.com/640-KB/GLaBIOS/files/12188524/GLABIOS_0.2.5_8F_072723_2.zip)

For some reason, that Multi Floppy BIOS is still loading twice.  Would you mind giving this BIOS a shot and see if this gives a clue?

**Comment by xinteksik** [comment]
> @xinteksik Oops, looks like the build I sent didn't have the option ROM display enabled after all. This one is the right one. :) [GLABIOS_0.2.5_8F_072723_2.zip](https://github.com/640-KB/GLaBIOS/files/12188524/GLABIOS_0.2.5_8F_072723_2.zip)
> 
> For some reason, that Multi Floppy BIOS is still loading twice. Would you mind giving this BIOS a shot and see if this gives a clue?

![IMG_1988](https://github.com/640-KB/GLaBIOS/assets/12784037/2b9a4dd7-af23-4b39-96c9-57fa6751098d)

I don´t know, why is loaded twice, becouse fdc bios (8k) is only one in ROM chip :)

UPDATE: Original ROM chip is 32K. I am using 64K chip. Maybe this is a problem of double loading - two different addres on bus are pointed to one space in ROM. I could try a 32K chip size. Maybe my opinion is wrong :)

**Comment by 640-KB** [comment]
> UPDATE: Original ROM chip is 32K. I am using 64K chip. Maybe this is a problem of double loading - two different addres on bus are pointed to one space in ROM. I could try a 32K chip size. Maybe my opinion is wrong :)

That could explain it.  If you have the Lo-Tech ROM card configured for a 64K, with a 32K ROM it would "alias" that 32K ROM twice into the space so it would do exactly what you are seeing where you get the same ROM contents mirrored 32K apart (`F400` + 32K = `FC00`).  Somehow that's what's going on since you're definitely seeing the high 32K of that ROM image appearing twice in the `F000` address space.  If you can get it to boot, you could confirm that by going to DEBUG and typing `d F600:0` and I bet you'll see GLaBIOS.

So yeah, I'd say experiment with different ROM chip sizes or DIP switch settings on that ROM card.  I'm betting that'll fix it... and if not it'll get one step closer -- I'm pretty sure that the floppy BIOS will not work properly if it's loaded twice.  

In the meantime, if you want to try - here's a build of GLaBIOS that only scans for option ROMs up to `F800` so it won't load the second instance of it.  This is of course only a hack to work around the core issue, but would at least tell you the ROM size is the problem and that your floppy BIOS is working!

[GLABIOS_0.2.5_8F_072823_1.zip](https://github.com/640-KB/GLaBIOS/files/12195737/GLABIOS_0.2.5_8F_072823_1.zip)

**Comment by xinteksik** [comment]
I can confirm that the latest BIOS loads the floppy BIOS only once. 

Btw the original ROM chip on The board is 32k and even if I used the same size or smaller 16k it always loads the BIOS twice. 

The PC with the last BIOS and floppy BIOS then booted from the HDD. Even if I set the size of the floppy drive correctly to 1.2M then the floppy can only be formatted to 360K.

![image](https://github.com/640-KB/GLaBIOS/assets/12784037/941ec1b2-175b-445e-8375-06e8c947b145)

**Comment by 640-KB** [comment]
@xinteksik are you using that Lo Tech ROM card at all right now or just the E/EPROM in the socket on the board?  If just the board ROM and you are seeing those images below `F800` then it sounds like that socket is wired for 64K (`F000-FFFF`).  I've seen this before where original BIOSes didn't look for option ROMs in the `F000` segment so even if it doubles it, nothing really happens (essentially what we did sending you that last BIOS image).

Do you have a 64K EPROM to try in the MB, and if so is there a reason you need the Lo Tech at all?  It feels like you *may* have two separate issues is all.

Does the now floppy drive read/write/boot, but only not format as 1.2M?  That screenshot looks like the drive type is right, so that should be okay.  If you go into Norton SI or CheckIt or something does it report the drive as being a 1.2 or a 360?  And you're 100% confident the drive, controller and disk you are testing are all totally good?

I've never used the DTK or the other floppy BIOS before myself, but I've used that [2M-XBIOS](https://www.[minuszerodegrees.net](https://www.minuszerodegrees.net/2M-XBIOS/2M-XBIOS%20-%201.44M%20as%20B.htm)/2M-XBIOS/2M-XBIOS%20-%201.44M%20as%20B.htm) TSR and has always worked really well for me. It might be interesting to try that (will want to remove the Multi-Floppy BIOS from the ROM image) and if it works we can eliminate anything related to hardware and focus on BIOS ROMs.

**Comment by xinteksik** [comment]
I have only EPROM in socket on motherboard. Original EPROM is 32k.

When i used only GLaBIOS with 2m-xbios in config.sys from HDD it works fine. I can use full capacity of floppy disk (1.2M). Read/write and format with "format a: /f:1.2".

When I used GLaBIOS with Multi-Floppy then read/write and format works but max 360KB. Hmm, previously formatted floppy disk (1.2M) is partially readable with this setup.

Board photo:
![image](https://github.com/640-KB/GLaBIOS/assets/12784037/ff460b65-5def-46d0-86b8-b95e430978d7)

![image](https://github.com/640-KB/GLaBIOS/assets/12784037/ca68bcc5-c61e-4256-a3dd-e3d25271abb2)

I don't have any manual or jumper setting for this board.

**Comment by 640-KB** [comment]
@xinteksik That's a pretty neat motherboard!  So, if the 2M-XBIOS works we can rule out hardware.  I don't think there will be any jumpers or switches relevant in this case - even if there were some special ones on this board for floppy drive configuration the Multi-Floppy BIOS doesn't read them and will just go with what's in the it's ROM.  Not really sure why the Multi-Floppy wouldn't like that motherboard, but perhaps just another will work better.

And with the Multi-Floppy BIOS, when you say it's partially readable, you can read and write to a 1.2M but still cannot format?  What do they say if you omit the parameters and just do a `format a: /u`?  Am also curious to know what DOS reports as the size of the floppy drive (like from SYSINFO or CheckIt), and if it's the same for both BIOSes, since the disk type and formatting capacity is handled pretty much entirely by DOS.

**Comment by xinteksik** [comment]
Well, some pictures.

Original ROM image. Everything OK.
CheckIT: A:1.2M
Max capacity: 1.2M

![IMG_2026](https://github.com/640-KB/GLaBIOS/assets/12784037/08fe64db-859b-4c7f-a937-415ab96e0254)
![IMG_2027](https://github.com/640-KB/GLaBIOS/assets/12784037/db02fa58-5958-417f-aabe-2bb12d3b6075)

Glabios ROM image + Multi-Floppy.  
CheckIT: A:Present (no size)
Max capacity: 360K

![IMG_2028](https://github.com/640-KB/GLaBIOS/assets/12784037/0b650334-461e-4974-8caa-c0236880f95d)
![IMG_2029](https://github.com/640-KB/GLaBIOS/assets/12784037/1d8340c0-5bff-4f89-91e3-c5128dfe50b2)

Glabios ROM image only + 2M-XBIOS in config.sys. Everything OK.
CheckIT: A:1.2M
Max capacity: 1.2M

![IMG_2033](https://github.com/640-KB/GLaBIOS/assets/12784037/615146d3-9219-4fd7-a9d7-d8976f2dde51)
![IMG_2034](https://github.com/640-KB/GLaBIOS/assets/12784037/2e77ee12-755a-4c64-903c-8a63251991ee)


Some points:
With Glabios ROM only, checkit detect size of floppy drive max 360K, but when i used in cmd "dir A:" nothing is happening (no spin drive) and get error.
With Glabios ROM and Multi-Floppy ROM, checkit detect size of floppy drive max 360K, but when i used in cmd "dir A:" i can see files on floppy disk. Some files I can copy, but some not.
It looks like geometry is from multifloppy but size form glabios :)

**Comment by 640-KB** [comment]
> Some points: With Glabios ROM only, checkit detect size of floppy drive max 360K, but when i used in cmd "dir A:" nothing is happening (no spin drive) and get error. With Glabios ROM and Multi-Floppy ROM, checkit detect size of floppy drive max 360K, but when i used in cmd "dir A:" i can see files on floppy disk. Some files I can copy, but some not. It looks like geometry is from multifloppy but size form glabios :)

So with GLaBIOS only, having it show as "present" (or technically what they call "not established") is what is expected since it doesn't support high density, or different drive types.  Though once a floppy disk ROM is installed, it completely takes over all floppy functions and data so GLaBIOS is totally out of the picture and plays no part in it.  

The big issue with XT's and high density floppy controllers/BIOSes is that there's no standard way to tell the PC/controller/BIOS the types of drives that are installed.  I believe a lot of those HD/BIOS controllers had jumper/DIP switches to select them, and if those don't exist/can't be read likely default to a 1.4M drive. That's the advantage of the Multi-Floppy BIOS is that it allows you to just hex edit and set the drive configuration.  However, what seems to be happening though with the Multi-Floppy ROM is that it's just actually not installing/initializing itself partially or completely.  

Here's something I wrote to help diagnose an issue with someone else's motherboard that gave us some useful clues and could help confirm my theory.  This program will dump the contents of memory to a file which if you want to do and upload it, I'll take a look.  This will give me the contents and configuration set up by Multi-Floppy which I can use to see if I can replicate what you see.  
https://github.com/rodneyknaap/atx-turboxt-v3/files/11962307/MEMDUMP.zip

**Comment by 640-KB** [comment]
One other troubleshooting idea...  could you try removing that Seagate hard drive controller and see if the Multi-Floppy BIOS works with the 1.2M drive?  Might give a clue...

**Comment by xinteksik** [comment]
> One other troubleshooting idea...  could you try removing that Seagate hard drive controller and see if the Multi-Floppy BIOS works with the 1.2M drive?  Might give a clue...

Heureka 😀 This works. 

With hdd controller (no hdd connected) - working.
![image](https://github.com/640-KB/GLaBIOS/assets/12784037/f3716489-cfdb-41b0-af86-8f2a903c5556)

With hdd controller and hdd - not working (partially read but after few seconds floppy can't read and computer hangs.
![image](https://github.com/640-KB/GLaBIOS/assets/12784037/0a6312b0-a91b-453c-a718-b10e4997271a)

**Comment by 640-KB** [comment]
> Heureka 😀 This works.
> 
> Maybe is problem in loading sequency? HDD ROM before floppy ROM?

Okay, this is excellent - and I was able to reproduce what you are seeing in 86Box. I also tried the DTK floppy BIOS and have the same problem... it seems like that Seagate 2.0.0 BIOS is doing "something bad" after the floppy BIOS initializes that makes it appear to DOS that the drive/controller is low density only.  

Interestingly, it works with the [ST11 ver 1.7 BIOS](https://github.com/86Box/roms/tree/master/hdd/st506) but not the 2.0.0 version.  Any chance the ROM on your board is socketed and could be swapped for the older version easily?

I tried swapping the order of the two ROMs so that the multi-floppy was first, but was never able to get the Seagate to work with the BIOS at any other address other than C800. Might be an 86Box issue though. It's worth a shot?

**Comment by xinteksik** [comment]
ROM is not in socket :( 

Just an idea, it is possible to disable load ROM from C000 (ROM on HDD controller) and loaded version 1.7 from ROM where is GLaBIOS and dual floppy?

![image](https://github.com/640-KB/GLaBIOS/assets/12784037/cccaf211-5546-4977-bd5d-08e64c5aa729)

**Comment by 640-KB** [comment]
You can change the BIOS address on the card from one of the 4 pre-defined options, which also changes the port I/O address as well.  Found this manual that describes the settings: https://www.lo-tech.co.uk/wiki/Seagate_ST11M_Installation_Guide

So in 86Box, by putting the Seagate's address at `D800` and port at `328` and then the Multi-Floppy BIOS at `D000` they work together.  Meaning, yes - if you load Multi-Floppy first, they seem to work. This would explain why the original BIOS that has the HD support built in does work too, since whatever bug in that Seagate ROM gets tripped if another storage controller is loaded *after* it.

Now doing that may become a little bit more tricky since the ROM sockets on the mother board are hard wired for `F000` which is way after that.  So you'd have to remove the Multi Floppy ROM from the EPROM on the board, use the Lo-Tech ROM adapter, and configure it for `D000-D800` and put the MF ROM there.  But worth a shot!

**Comment by xinteksik** [comment]
Thanks. But I don't have a Lo-Tech ROM adapter 😭.

**Comment by 640-KB** [comment]
@xinteksik 
So here's a "hacked" version that alters the ROM scan a little bit.  It first scans `F000`-`F800` and then `C800`-`F000` so the ROMs in the motherboard EPROM will be read first.  This way it will load the Multi-Floppy first, then go back to the Seagate BIOS which (if this works) will workaround the bug.

[GLABIOS_0.2.5_8F_X1.zip](https://github.com/640-KB/GLaBIOS/files/12231724/GLABIOS_0.2.5_8F_X1.zip)

Fingers crossed!

**Reply by Batsch69** [reply]
I have a similar problem. IBM PC XT 5160 v2/3, Seagate ST11 V2.0 BIOS + Floppy drive ROM at F200.
Seagte is loaded first, then FloppyBios. FloppyBios hangs.
Is there a option in "GLABIOS.ASM" to first load F000-F800 and then C800-F000?

**Comment by xinteksik** [comment]
Thanks a lot! I gues mission is accomplished 👍

I used only Glabios and Dual Floppy. 

![image](https://github.com/640-KB/GLaBIOS/assets/12784037/4675d7bb-bd86-4618-bcb2-47c5c980b16f)

![image](https://github.com/640-KB/GLaBIOS/assets/12784037/f5142be3-4614-462b-bc73-62d437b0aa34)

![image](https://github.com/640-KB/GLaBIOS/assets/12784037/49630440-0138-4761-91fc-08bde34ed827)

**Comment by 640-KB** [comment, marked as answer]
@xinteksik Wow! What a mystery to solve! We've uncovered yet another ancient bug that might have been forever forgotten otherwise.  This should go into the [GLaBIOS wiki](https://github.com/640-KB/GLaBIOS/wiki) so that someday it might help someone else.

I guess just let me know if you run into any other problems or have any thoughts or suggestions!

640KB
